src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java

```diff
@@ -595,27 +612,60 @@ private void refreshTextureCache()
         final int displayPerSlots = this.getInventory().getSlots() / 4;
         int index = 0;
         boolean update = false;
+        boolean alreadyAddedItem = false;
+
+        final HashMap<ItemStorage, Integer> mapCopy = new HashMap<>(content);
+        if (this.getOtherChest() instanceof TileEntityRack neighborRack)
+        {
+            for (final Map.Entry<ItemStorage, Integer> entry : neighborRack.content.entrySet())
+            {
+                int value = entry.getValue() + mapCopy.getOrDefault(entry.getKey(), 0);
+                mapCopy.put(entry.getKey(), value);
+            }
+        }
+        final List<Map.Entry<ItemStorage, Integer>> list = mapCopy.entrySet().stream().sorted((e1, e2) -> Integer.compare(e2.getValue(), e1.getValue())).toList();
 
-        final List<Map.Entry<ItemStorage, Integer>> list = content.entrySet().stream().sorted((e1, e2) -> Integer.compare(e2.getValue(), e1.getValue())).toList();
         final Queue<Block> extraBlockQueue = new ArrayDeque<>();
+        final Queue<Block> itemQueue = new ArrayDeque<>();
         for (final Map.Entry<ItemStorage, Integer> entry : list)
         {
             // Need more solid checks!
             if (index < textureMapping.size())
             {
                 Block block = Blocks.BARREL;
-                if (entry.getKey().getItemStack().getItem() instanceof BlockItem blockitem
-                      && blockitem.getBlock().defaultBlockState().isSolidRender(EmptyBlockGetter.INSTANCE, BlockPos.ZERO))
+                boolean isBlockItem = false;
+                if (entry.getKey().getItemStack().getItem() instanceof BlockItem blockItem)
                 {
-                    block = blockitem.getBlock();
+                    block = blockItem.getBlock();
+                    isBlockItem = true;
                 }
 
                 int displayRows = (int) Math.ceil((Math.max(1.0, (double) entry.getValue() / entry.getKey().getItemStack().getMaxStackSize())) / displayPerSlots);
                 if (displayRows > 1)
                 {
                     for (int i = 0; i < displayRows - 1; i++)
                     {
-                        extraBlockQueue.add(block);
+                        if (isBlockItem)
+                        {
+                            extraBlockQueue.add(block);
+                        }
+                        else
+                        {
+                            itemQueue.add(block);
+                        }
+                    }
+                }
+
+                if (!isBlockItem)
+                {
+                    if (alreadyAddedItem)
```

Comment:
Did you miss a "!" here?

Reply:
Not, itentional. If we already added it, we stop here and just add it to the queue, else we pass and mark true